### PR TITLE
[ROCm][Perf] Enable W4A16 FlyDSL MoE

### PR DESCRIPTION
**Pull request by amd-asalykov**
## Purpose
This PR replaces w4a16 MoE GPTQ Triton kernel with custom FlyDSL w4a16 MoE kernel

1. weights are int 4-bit quantized
2. group_size==32
3. arch==gfx950 (currently limited to gfx950, as the MoE was tuned only for gfx950)
4. --moe-backend=flydsl
5. lora is disabled

targeting Kimi K2.5 INT4 model. The kernel differs from `aiter.fused_moe` implementation, as we pre-shuffle/flatten weights+scales during weight initialization instead at inference time to increase the perf.

## Test Plan

Measure e2e perf. and test accuracy on gsm8k

## Test Result
Recommended command (currently, enabling fuse_allreduce_rms results in accuracy=0, therefore disabling it)
```bash
export VLLM_ROCM_USE_AITER=1

vllm serve moonshotai/Kimi-K2.5 \
  --tensor-parallel-size=8 --gpu-memory-utilization 0.95 \
  --block-size=64 --trust-remote-code \
  --mm-encoder-tp-mode data \
  --moe-backend flydsl \
  --compilation-config '{"pass_config": {"fuse_allreduce_rms": false}}'
```

Accuracy: ~0.934 on gsm8k task
Perf. MI355X:

8k1kpr
1k1kpr

### CODE REVIEW

**Comment by tjtanaa** [comment]
@amd-asalykov are you using the flydsl version in the `vllm/vllm-openai-rocm:nightly`? I think it is `flydsl==0.1.4.2`. Does this kernels work with flydsl `0.1.4.2`?

File: vllm/envs.py [hunk 1]
```diff
@@ -1102,6 +1102,11 @@ def _resolve_rust_frontend_path() -> str | None:
     "VLLM_ROCM_USE_AITER_MOE": lambda: (
         os.getenv("VLLM_ROCM_USE_AITER_MOE", "True").lower() in ("true", "1")
     ),
+    # Whether to use flydsl moe op.
```
Comment (tjtanaa) [inline vllm/envs.py]: For MOE, we do not rely on environment flag.

We should use the python argument interface instead

`--moe-backend` https://docs.vllm.ai/en/latest/cli/serve/#-moe-backend

you can name is `flydsl`

File: vllm/model_executor/layers/fused_moe/fused_flydsl_moe.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tjtanaa) [inline vllm/model_executor/layers/fused_moe/fused_flydsl_moe.py:1]: Is this kernel generalised to all type of input shapes and expert parallelism configuration?
This question will help us to understand if we should set it as a default for gfx950.

File: vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json [hunk 3]
```diff
@@ -0,0 +1,114 @@
+{
```
Comment (tjtanaa) [inline vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: @amd-asalykov please include the tuning script you used to generate these `.json` file in `benchmarks/kernels`.

File: tests/kernels/moe/test_flydsl_moe.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    # FlyDSL interleaved int4 packing
+    packed = _pack_shuffled_int8_to_packed_int4_no_perm(shuffled).contiguous()
+    return packed.view(E, N, K // 2)
+
+
+@pytest.mark.skipif(
+    not (current_platform.is_rocm() and on_gfx950()),
+    reason="FlyDSL MoE requires HIP device and gfx950 arch",
+)
+@pytest.mark.parametrize("num_tokens", [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024])
```
Comment (tjtanaa) [inline tests/kernels/moe/test_flydsl_moe.py]: for the token size please add a test till `16384`. This is the typical `max-num-batched-token` size.

Offline inference `max-num-batched-token` is `16384`
Online inference `max-num-batched-token` is `8192`

**Comment by dllehr-amd** [comment]
@divakar-amd  can you take a quick look as well?

File: vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json (hunk 3 above)
Comment (rasmith) [inline vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: Can these tuning files be stored in AITER instead?

Comment (tjtanaa) [reply vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: @rasmith 

Primary reason: Since this kernel is written directly in vLLM, we should keep the tuning script and config.json in vLLM. 

Secondary reason: There is a very long lag between aiter and vllm because we are not able to upgrade aiter version whenever we need to. That is why our vLLM out of the box is always lagging behind in performance because we cannot get any of the new tuning config from AITER as we cannot upgrade AITER.

CC @dllehr-amd @AndreasKaratzas 

Users and developers can tune the kernel for new models and get the performance gain immediately.

Comment (tjtanaa) [reply vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: Moreover, AITER is not a repo to store all sorts of user space tuning config. It should be a repo that gives downstream users the kernels and a tuning system, then users tune the kernels for their own Model and Deployment configuration.

So, when this issue https://github.com/ROCm/aiter/issues/2797 is resolved, we should also store the aiter's config in vLLM instead of AITER because:
1. The compatibility of the tuning config is tied to an AITER version. AITER tuning config is always aligns with their latest features. Downstream user of AITER usually pinned to a specific AITER version like how vLLM managed the AITER version
2. Downstream user of AITER like vLLM has their own TP/DP/EP/PP implementation. So the configuration is not suitable to be stored in AITER. (But I believe from AMD perspective, you guys can still upstream the tuning config back to AITER, but as AITER user, we should keep them in user workspace )
3. The lag between AITER version upgrade as each downstream user of AITER has their own upgrade cadence.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        if self.moe.is_lora_enabled:
+            assert self.moe_quant_config is not None
+            from vllm.triton_utils import HAS_TRITON
+
+            if HAS_TRITON:
+                from vllm.model_executor.layers.fused_moe import TritonWNA16Experts
+
+                layer.w13_weight = layer.w13_weight_packed
+                layer.w2_weight = layer.w2_weight_packed
+                return TritonWNA16Experts(
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py]: process_weights_after_loading runs before select_gemm_impl. By the time this branch runs, process_weights_after_loading has already done w13.view(-1) / w2.view(-1), so the packed tensors are flat 1-D blobs in the FlyDSL packed-int4 byte layout not the [E, K//8, N] int32 GPTQ layout that TritonWNA16Experts expects. The Triton fallback will either crash or silently produce garbage.

Comment (amd-asalykov) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py]: @okorzh-amd thank you very much for the review! When running Kimi int4 model, I observed that the function is never called and therefore didn't patch it. select_gemm_impl branch will only be executed if lora is enabled in moe. I can add check in `compressed_tensors_moe.py‎` for lora and route to default triton w4a16 moe implementation if lora is enabled

Comment (okorzh-amd) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py]: Thanks, that matches what I suspected (the branch is only reachable when LoRA is enabled on the MoE layer, which Kimi K2 deployments don't hit). Routing to the default Triton W4A16 path in compressed_tensors_moe.py when LoRA is enabled is the cleaner of the two options I had in mind, since it preserves a working W4A16+LoRA path instead of just hard-erroring at construction time.

File: vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json (hunk 3 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: That is actually true, and it also aligns with the current efforts by the Helion team to introduce helion kernels. So I am not against that.

Comment (salykova) [reply vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json]: [@tjtanaa](https://github.com/tjtanaa) [@dllehr-amd](https://github.com/dllehr-amd) [@AndreasKaratzas](https://github.com/AndreasKaratzas) [@rasmith](https://github.com/rasmith)  I agree with [@tjtanaa](https://github.com/tjtanaa) that the user space tuning configs should be stored in vllm, rather than in AITER to avoid something like this on the user side  https://github.com/vllm-project/vllm/pull/39987

image

File: vllm/envs.py (hunk 1 above)
Comment (amd-asalykov) [reply vllm/envs.py]: @tjtanaa fixed

File: vllm/model_executor/layers/fused_moe/fused_flydsl_moe.py (hunk 2 above)
Comment (amd-asalykov) [reply vllm/model_executor/layers/fused_moe/fused_flydsl_moe.py:1]: this kernel generalizes to arbitrary shapes, but doesn't support expert parallelism afaik

File: vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json (hunk 3 above)
Comment (amd-asalykov) [reply vllm/model_executor/layers/fused_moe/configs/E=384,N=256,device_name=AMD_Instinct_MI350_OAM,dtype=int4_w4a16,backend=flydsl.json:1]: @tjtanaa done

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py (hunk 5 above)
Comment (amd-asalykov) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a16_flydsl.py]: @okorzh-amd fixed

**Comment by tjtanaa** [comment]
@amd-asalykov can you revert the `vllm/envs.py` file mode? it still showed up as a diff.
image

**Comment by amd-asalykov** [comment]
@tjtanaa I reverted the `vllm/envs.py` file mode. Could you please help to proceed with the PR?

**Comment by tjtanaa** [comment]
@amd-asalykov 

(the other arguments on top of your command are from inferenceX) In other words, I add `--compilation-config '{"pass_config": {"fuse_allreduce_rms": false}}'` and `--moe-backend flydsl \` to InferenceX command.
```
#!/bin/bash
rm -rf ~/.cache/vllm

export VLLM_ROCM_USE_AITER=1

vllm serve moonshotai/Kimi-K2.5 \
--port 8894 \
--tensor-parallel-size=8 \
--gpu-memory-utilization 0.95 \
--block-size=64 \
--trust-remote-code \
--no-enable-prefix-caching \
--max-num-seqs 256 \
--mm-encoder-tp-mode data \
--moe-backend flydsl \
--enable-expert-parallel \
--compilation-config '{"pass_config": {"fuse_allreduce_rms": false}}'
```

I have done a simple probe before running full gsm8k dataset, it seems EP is giving gibberish answer

```
/app/reviewpr06/flydsl-w4a16-moe# python3 examples/basic/online_serving/openai_chat_completion_client.py 
--------------------------------------------------
Chat completion results:
ChatCompletion(id='chatcmpl-b32e4c7bde9fcf9b', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content=' ;\n\n confidence $ aims options sources rumoursjournalisty variables Put integersapters hilarious newspaper over choices imported variations move', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=[], reasoning=None), stop_reason=None, token_ids=None, routed_experts=None)], created=1781323681, model='moonshotai/Kimi-K2.5', object='chat.completion', moderation=None, service_tier=None, system_fingerprint='vllm-0.19.1rc1.dev2085+gf44a19c35-tp8-ep-3fb8624c', usage=CompletionUsage(completion_tokens=20, prompt_tokens=56, total_tokens=76, completion_tokens_details=None, prompt_tokens_details=None), prompt_logprobs=None, prompt_token_ids=None, prompt_text=None, kv_transfer_params=None)
```


With your command without EP, it looks fine

```
/app/reviewpr06/flydsl-w4a16-moe# python3 examples/basic/online_serving/openai_chat_completion_client.py 
--------------------------------------------------
Chat completion results:
ChatCompletion(id='chatcmpl-a607fa7474611976', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content=' The user is asking where the 2020 World Series was played. I need to recall that the', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=[], reasoning=None), stop_reason=None, token_ids=None, routed_experts=None)], created=1781324031, model='moonshotai/Kimi-K2.5', object='chat.completion', moderation=None, service_tier=None, system_fingerprint='vllm-0.19.1rc1.dev2085+gf44a19c35-tp8-752bd573', usage=CompletionUsage(completion_tokens=20, prompt_tokens=56, total_tokens=76, completion_tokens_details=None, prompt_tokens_details=None), prompt_logprobs=None, prompt_token_ids=None, prompt_text=None, kv_transfer_params=None)
--------------------------------------------------
```

**Comment by amd-asalykov** [comment]
@tjtanaa yes, the FlyDSL kernel doesn't support EP. InferenceX uses EP only for FP4 model. INT4 runs without EP [kimik2.5_int4_mi355x.sh](https://github.com/SemiAnalysisAI/InferenceX/blob/main/benchmarks/single_node/fixed_seq_len/kimik2.5_int4_mi355x.sh) [kimik2.5_fp4_mi355x.sh](https://github.com/SemiAnalysisAI/InferenceX/blob/main/benchmarks/single_node/fixed_seq_len/kimik2.5_fp4_mi355x.sh)

I also noticed that 3 CI tests failed, but I checked the logs and the errors should be unrelated to the changes in this PR if I understood it correctly

**Comment by tjtanaa** [comment]
@amd-asalykov will there be plans to follow up on supporting EP?

**Review by tjtanaa — APPROVED** [review]
LGTM. Will have to update the vllm recipe next.

Also did a cross validation without EP

```
#!/bin/bash
rm -rf ~/.cache/vllm

export VLLM_ROCM_USE_AITER=1

vllm serve moonshotai/Kimi-K2.5 \
--port 8894 \
--tensor-parallel-size=8 \
--gpu-memory-utilization 0.95 \
--block-size=64 \
--trust-remote-code \
--no-enable-prefix-caching \
--max-num-seqs 256 \
--mm-encoder-tp-mode data \
--moe-backend flydsl \
--compilation-config '{"pass_config": {"fuse_allreduce_rms": false}}' \
```

GSM8K with large concurrrency and longer inputs

```
local-completions ({'model': 'moonshotai/Kimi-K2.5', 'base_url': 'http://0.0.0.0:8894/v1/completions', 'num_concurrent': 256, 'max_retries': 10, 'max_gen_toks': 2048, 'max_length': 1048576, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 20, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|_  |0.9386|_  |0.0066|
|     |       |strict-match    |    20|exact_match|_  |0.9393|_  |0.0066|
```

**Comment by tjtanaa** [comment]
> @tjtanaa yes, the FlyDSL kernel doesn't support EP. InferenceX uses EP only for FP4 model. INT4 runs without EP [kimik2.5_int4_mi355x.sh](https://github.com/SemiAnalysisAI/InferenceX/blob/main/benchmarks/single_node/fixed_seq_len/kimik2.5_int4_mi355x.sh) [kimik2.5_fp4_mi355x.sh](https://github.com/SemiAnalysisAI/InferenceX/blob/main/benchmarks/single_node/fixed_seq_len/kimik2.5_fp4_mi355x.sh)
> 
> I also noticed that 3 CI tests failed, but I checked the logs and the errors should be unrelated to the changes in this PR if I understood it correctly

I am retrying the tests, they could be flaky.

**Comment by amd-asalykov** [comment]
@tjtanaa hmm one test still fails, but it seems the errors are unrelated to the PR:

[buildkite/ci/pr/amd-v1-sample-plus-logits-mi325-1](https://buildkite.com/vllm/ci/builds/71926#019ec16b-4c78-4a04-8f7f-e532392e48cc)

image

**Review by fxmarty-amd — COMMENTED** [review]
@tjtanaa @mgoin we should not merge PRs that do not make use of the oracle abstraction for linear/MOE backends.

This PR does not work with other model producers than compressed-tensors

**Comment by fxmarty-amd** [comment]
@amd-asalykov for reference:
- https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/oracle/mxfp4.py
- https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/oracle/mxfp4.py
